### PR TITLE
Wangc/bison interface

### LINKED ISSUE #1609: [DEFECT] MooseBasedApp Interface error with moose [Outputs] Block
--------
Defect Description
--------

When using the MooseBasedApp interface in raven, in the [Outputs] block of a moose based input deck, if a `[csv]` type block is already used, the addition of the line `csv=true` to the `[Outputs]` block results in an error in the moose based app. 

##### What did you expect to see happen?
In the original input deck the use of [no_print] block allows the printing of the output from the moose simulation.

##### What did you see instead?
The RAVEN modified input deck adds the line `csv=true` and `file_base=out~input` to the `[Outputs]` block. The inclusion of `csv=true`  results in an error stating the `file_base` input already exists and stops the run. 

##### Do you have a suggested fix for the development team?
If the `[csv]` type block is already included in the `[Outputs]` block then the line `csv=true` does not need to be added to the input deck by the interface.

**Describe how to Reproduce**
1.  In a MooseBasedApp input deck include the [csv] block under [Outputs]
2.  Run RAVEN case with MooseBasedApp interface
 
**Screenshots and Input Files**

Original Input Deck:
```
[Outputs]
# Define output file(s)
  interval = 1
  print_linear_residuals = true
  perf_graph = true
  color = false

  [console]
    type = Console
    max_rows = 25
  []

  [exodus]
    type = Exodus
    interval = 3
  []

  [no_print]
    type = csv
  []
[]

```

Input deck modified by RAVEN:
```
[Outputs]
  interval = 1
  print_linear_residuals = true
  perf_graph = true
  color = false
  [./console]
    type = Console
    max_rows = 25
  [../]
  [./exodus]
    type = Exodus
    interval = 3
  [../]
  [./no_print]
     type=csv
  [../]
  csv = true
  file_base = out~input
[]
```

**Platform (please complete the following information):**
 - Version: [e.g. 22]
 - Dependencies Installation: [CONDA or PIP]

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #2080: Process Failed <ravenframework.Runners.SharedMemoryRunner.SharedMemoryRunner
--------
Issue Description
--------
##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

[Errorlog.txt](https://github.com/idaholab/raven/files/10801937/Errorlog.txt)
[XMLfile.txt](https://github.com/idaholab/raven/files/10801942/XMLfile.txt)

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [ ] 1. Is it tagged with a type: defect or task?
- [ ] 2. Is it tagged with a priority: critical, normal or minor?
- [ ] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [ ] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [ ] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [ ] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #2159: [TASK] Support top level variables in moose inputs
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
MOOSE supports inputs to live outside of typical `[]` tags like in this [arbitrarily chosen test](https://github.com/idaholab/moose/blob/98c121565021040d042a982ab2ca0726c512b7b2/modules/thermal_hydraulics/test/tests/components/hs_coupler_2d3d/hs_coupler_2d3d.i#L3-L9C13). Using such an input in raven cause it to fail. 

```
  File "raven/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py", line 138, in getpotToInputTree
    if attribute in (c.tag for c in currentNode.children): #currentNode.attrib:
AttributeError: 'NoneType' object has no attribute 'children'
```

**Describe the solution you'd like**
I'd like to see RAVEN be able to handle this typical moose input. 

Here is a patch that would do  but I have no idea if I've missed some gotcha's, like at the least this doesn't check for a duplicate anymore on a top level input. 

```
diff --git a/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py b/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py
index e2edf54b1..f5889c1bd 100644
--- a/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py
+++ b/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py
@@ -135,7 +135,11 @@ def getpotToInputTree(getpot):
     #------------------
     # if the "value" if the "attribute" is closed, record the entry
     if closeEntry:
-      if attribute in (c.tag for c in currentNode.children): #currentNode.attrib:
+      if currentNode is None:
+        currentNode = TreeStructure.InputNode(tag=attribute, text=value)
+        roots.append(currentNode)
+        currentNode = None
+      elif attribute in (c.tag for c in currentNode.children): #currentNode.attrib:
         raise IOError('Multiple entries defined with same name "{a}"!'.format(a=attribute))
       else:
         new = TreeStructure.InputNode(tag=attribute, text=value)
```

and this allows the var to be sampled:

```
diff --git a/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseBasedAppInterface.py b/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseBasedAppInterface.py
index 8f7fcb11d..857c1137f 100644
--- a/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseBasedAppInterface.py
+++ b/ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseBasedAppInterface.py
@@ -160,9 +160,6 @@ class MooseBasedApp(CodeInterfaceBase):
         request, elemLoc = (v.strip() for v in var.split(':'))
       else:
         request = var
-      if '|' not in request:
-        # what modifications don't have the path in them?
-        continue
       pathedName = request.split('|')
       modifDict['name'] = pathedName
       if elemLoc is not None:
```

**Describe alternatives you've considered**
Don't use top level inputs, but they are just too convenient. At the least a nice error when this is detected would be nice.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
This PR will address the following issues:

close #2159 
close #1609 
close #2080 

##### What are the significant changes in functionality due to this change request?

Allow moose interface to accept top level variables, as well as some improvements on how to output csv file. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by wangcj05** [comment]
Hi @PaulTalbot-INL and  @joshua-cogliati-inl, can any of you have time to review this PR?

File: ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py [hunk 1]
```diff
@@ -255,6 +259,45 @@ def findInGetpot(trees, targetPath):
         break
   return found
 
+def findCsvInGetpot(trees):
+  """
+    Find the CSV output blocks in the input file
+    @ In, trees, list(TreeStructure.InputTrees), the tree(s) represent the input file
+    @ Out, list, list(TreeStructure.InputTrees), list of CSV blocks in the input file
```
Comment (PaulTalbot-INL) [inline ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py:266]: this looks like it's returning a bool, not a list, right?

File: tests/framework/CodeInterfaceTests/MOOSEBaseApps/input_parser_csv.xml [hunk 2]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/CodeInterfaceTests/MOOSEBaseApps.InputParser</name>
+    <author>talbpaul</author>
+    <created>2020-04-06</created>
```
Comment (PaulTalbot-INL) [inline tests/framework/CodeInterfaceTests/MOOSEBaseApps/input_parser_csv.xml:6]: is this a file relocated or a new file? Update authorship and date if it's new.

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
Two small comments to consider.

File: ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py (hunk 1 above)
Comment (wangcj05) [reply ravenframework/CodeInterfaceClasses/MooseBasedApp/MooseInputParser.py]: Good catch, I have updated the docstring.

File: tests/framework/CodeInterfaceTests/MOOSEBaseApps/input_parser_csv.xml (hunk 2 above)
Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/MOOSEBaseApps/input_parser_csv.xml]: Updated.

**Comment by wangcj05** [comment]
@PaulTalbot-INL Thanks for your comments, I have resolved your comments.

**Review by PaulTalbot-INL — APPROVED** [review]

**Comment by PaulTalbot-INL** [comment]
Good to merge if tests pass without further changes.

**Pull request merged**